### PR TITLE
Animation: Tower and Magic Arrow

### PR DESCRIPTION
**Pull request by idshibanov**
Refactor to use the improved code as well as generalizing code in RedrawAttack to make it cleaner. Moved ICN selectors to `Monster` class. Angles, target position and projectile selection should be better. `RedrawActionTowerPart1` is a great example.

Also corrected the mage beam speed and added small OG delay between unit anim and beam anim.

### CODE REVIEW

File: src/fheroes2/battle/battle_interface.cpp [hunk 1]
```diff
@@ -2470,37 +2486,39 @@ void Battle::Interface::RedrawActionAttackPart1( Unit & attacker, Unit & defende
...
-        while ( le.HandleEvents( false ) && pnt != points.end() ) {
-            CheckGlobalEvents( le );
+            // Mage is chanelling the bolt
+            DELAY( Game::ApplyBattleSpeed( 115 ) );
 
-            if ( Battle::AnimateInfrequentDelay( Game::BATTLE_MISSILE_DELAY ) ) {
-                cursor.Hide();
-                Redraw();
-                if ( attacker.GetID() == Monster::MAGE || attacker.GetID() == Monster::ARCHMAGE ) {
+            while ( le.HandleEvents( false ) && pnt != points.end() ) {
```
Comment (ihhub) [inline src/fheroes2/battle/battle_interface.cpp:2501]: Could we please put mage shooting in `RedrawMissileAnimation` or in a separate method to have a clearer code? Logically, if we put all code for shooting in a separate method then we need to do the same for mages from design point of view.

File: src/fheroes2/battle/battle_interface.cpp [hunk 2]
```diff
@@ -2409,12 +2408,37 @@ void Battle::Interface::RedrawActionSkipStatus( const Unit & attacker )
     status.SetMessage( msg, true );
 }
 
-void Battle::Interface::RedrawActionAttackPart1( Unit & attacker, Unit & defender, const TargetsInfo & targets )
+void Battle::Interface::RedrawMissileAnimation( Point startPos, Point target, double angle, uint32_t monsterID )
```
Comment (ihhub) [inline src/fheroes2/battle/battle_interface.cpp:2411]: Please make `startPos` and `target` input variables as const references. Also it is good to renamed **target** variable into **endPos** to have a contrast named variables.

File: src/fheroes2/game/game.cpp [hunk 3]
```diff
@@ -429,10 +429,7 @@ void Game::LoadExternalResource( const Settings & conf )
         Game::UpdateGlobalDefines( spec );
 
     // animations.xml
-    spec = Settings::GetLastFile( prefix_stats, "animations.xml" );
-
-    if ( System::IsFile( spec ) )
-        Battle::UpdateMonsterSpriteAnimation( spec );
+    // spec = Settings::GetLastFile( prefix_stats, "animations.xml" );
```
Comment (ihhub) [inline src/fheroes2/game/game.cpp:432]: Should we remove this comment?

File: src/fheroes2/battle/battle_interface.cpp [hunk 4]
```diff
@@ -2409,12 +2408,37 @@ void Battle::Interface::RedrawActionSkipStatus( const Unit & attacker )
...
+    Points::const_iterator pnt = points.begin();
+
+    // convert the following code into a function/event service
+    while ( le.HandleEvents( false ) && pnt != points.end() ) {
+        CheckGlobalEvents( le );
+
+        if ( Battle::AnimateInfrequentDelay( Game::BATTLE_MISSILE_DELAY ) ) {
+            cursor.Hide();
+            Redraw();
+            missile.Blit( reverse ? pnt->x - missile.w() : pnt->x, ( angle > 0 ) ? pnt->y - missile.h() : pnt->y );
```
Comment (ihhub) [inline src/fheroes2/battle/battle_interface.cpp:2431]: For mages here we could do something like this:
```
// before the loop
const bool isMage = ( monsterID == Monster::MAGE || monsterID == Monster::ARCHMAGE );
...
// Mage is channeling the bolt <--- misspelled channeling :)
if ( isMage ) {
    DELAY( Game::ApplyBattleSpeed( 115 ) );
}
...
// in the loop
if ( isMage ) {
    display.DrawLine( Point( missileStart.x, missileStart.y - 2 ), Point( pnt->x, pnt->y - 2 ), PAL::GetPaletteColor( 0x77 ) );
    display.DrawLine( Point( missileStart.x, missileStart.y - 1 ), Point( pnt->x, pnt->y - 1 ), PAL::GetPaletteColor( 0xB5 ) );
    display.DrawLine( Point( missileStart.x, missileStart.y ), Point( pnt->x, pnt->y ), PAL::GetPaletteColor( 0xBC ) );
    display.DrawLine( Point( missileStart.x, missileStart.y + 1 ), Point( pnt->x, pnt->y + 1 ), PAL::GetPaletteColor( 0xB5 ) );
    display.DrawLine( Point( missileStart.x, missileStart.y + 2 ), Point( pnt->x, pnt->y + 2 ), PAL::GetPaletteColor( 0x77 ) );
}
else {
    missile.Blit( reverse ? pnt->x - missile.w() : pnt->x, ( angle > 0 ) ? pnt->y - missile.h() : pnt->y );
}
```
An extra boolean check is nothing for the performance. Please correct me if I am wrong.

File: src/fheroes2/battle/battle_interface.cpp [hunk 5]
```diff
@@ -2409,12 +2408,55 @@ void Battle::Interface::RedrawActionSkipStatus( const Unit & attacker )
...
+
+    const bool reverse = startPos.x > endPos.x;
+    const bool isMage = ( monsterID == Monster::MAGE || monsterID == Monster::ARCHMAGE );
+
+    // Mage is channeling the bolt
+    if ( isMage ) {
+        DELAY( Game::ApplyBattleSpeed( 115 ) );
+    }
+    else {
+        missile = AGG::GetICN( Monster::GetMissileICN( monsterID ), Bin_Info::GetMonsterInfo( monsterID ).getProjectileID( angle ), reverse );
```
Comment (idshibanov) [inline src/fheroes2/battle/battle_interface.cpp:2426]: Updated this as you suggested. However I removed druid missile mapping for mages, so if you'll try to get this ref game will crash. Have to handle it like that, are you fine with this? Or it's better to have a fallback for missile?

Comment (ihhub) [reply src/fheroes2/battle/battle_interface.cpp:2426]: I think if we try to get a missile for a Dragon for example the game crashes as well, doesn't it? :)

Comment (ihhub) [reply src/fheroes2/battle/battle_interface.cpp:2426]: It crashes in `LoadICN` function which an exception. We should not do this as it seems weird and lead to an immediate crash. What we could do: the function should return boolean variable. Then in `GetICN` method we will have such code:
```
if ( !LoadICN( icn, index, reflect ) ) {
    return result;
}
```

We should still have a logging about incorrect ICN but gracefully handle this situation.

Comment (idshibanov) [reply src/fheroes2/battle/battle_interface.cpp:2426]: Makes sense, I'll add the handling

**Comment by ihhub** [comment]
@idshibanov , thank you very much for your heave contribution to the project!

**Pull request merged**